library/init/category/reader.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+
+The reader monad transformer for passing immutable state.
+-/
+
+prelude
+import init.category.lift init.category.id
+universes u v w
+
+structure reader_t (r : Type u) (m : Type u → Type v) (α : Type u) : Type (max u v) :=
+(run : r → m α)
+
+@[reducible] def reader (r : Type u) := reader_t r id
+
+namespace reader_t
+section
+  variable  {r : Type u}
+  variable  {m : Type u → Type v}
+  variable  [monad m]
+  variables {α β : Type u}
+
+  protected def read : reader_t r m r :=
+  ⟨pure⟩
+
+  protected def pure (a : α) : reader_t r m α :=
+  ⟨λ r, pure a⟩
+
+  protected def bind (x : reader_t r m α) (f : α → reader_t r m β) : reader_t r m β :=
+  ⟨λ r, do a ← x.run r,
+           (f a).run r⟩
+
+  instance : monad (reader_t r m) :=
+  { pure := @reader_t.pure _ _ _, bind := @reader_t.bind _ _ _ }
+
+  protected def lift (a : m α) : reader_t r m α :=
+  ⟨λ r, a⟩
+
+  instance (m) [monad m] : has_monad_lift m (reader_t r m) :=
+  ⟨@reader_t.lift r m _⟩
+
+  protected def monad_map {r m m'} [monad m] [monad m'] {α} (f : Π {α}, m α → m' α) : reader_t r m α → reader_t r m' α :=
+  λ x, ⟨λ r, f (x.run r)⟩
+
+  instance (r m m') [monad m] [monad m'] : monad_functor m m' (reader_t r m) (reader_t r m') :=
```

Comment:
It is clear to me why you defined `has_monad_lift m (reader_t r m)`, but why do we need this `monad_functor` instance?